pallets/parachain-staking/src/inflation.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright 2020-2022 Manta Network.
+// This file is part of Manta.
+//
+// Manta is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Manta is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Manta.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Helper methods for computing issuance based on inflation
+use crate::pallet::{BalanceOf, Config, Pallet};
+use frame_support::traits::Currency;
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+use sp_runtime::{PerThing, Perbill, RuntimeDebug};
+use substrate_fixed::{
+    transcendental::pow as floatpow,
+    types::{I32F32, I64F64},
+};
+
+const SECONDS_PER_YEAR: u32 = 31557600; // = 365.25 * 24 * 60 * 60, .25 to accommodate leap years
+const SECONDS_PER_BLOCK: u32 = manta_primitives::constants::time::SECONDS_PER_BLOCK as u32;
+pub const BLOCKS_PER_YEAR: u32 = SECONDS_PER_YEAR / SECONDS_PER_BLOCK;
+
+fn rounds_per_year<T: Config>() -> u32 {
+    let blocks_per_round = <Pallet<T>>::round().length;
+    BLOCKS_PER_YEAR / blocks_per_round
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(
+    Eq, PartialEq, Clone, Copy, Encode, Decode, Default, RuntimeDebug, MaxEncodedLen, TypeInfo,
+)]
+pub struct Range<T> {
+    pub min: T,
+    pub ideal: T, // Target rate of inflation. round length will vary to keep issuance close to this setting
+    pub max: T,
+}
+
+impl<T: Ord> Range<T> {
+    pub fn is_valid(&self) -> bool {
+        self.max >= self.ideal && self.ideal >= self.min
+    }
+}
+
+impl<T: Ord + Copy> From<T> for Range<T> {
+    fn from(other: T) -> Range<T> {
+        Range {
+            min: other,
+            ideal: other,
+            max: other,
+        }
+    }
+}
+/// Convert an annual inflation to a round inflation
+/// round = (1+annual)^(1/rounds_per_year) - 1
+pub fn perbill_annual_to_perbill_round(
+    annual: Range<Perbill>,
+    rounds_per_year: u32,
+) -> Range<Perbill> {
+    let exponent = I32F32::from_num(1) / I32F32::from_num(rounds_per_year);
+    let annual_to_round = |annual: Perbill| -> Perbill {
+        let x = I32F32::from_num(annual.deconstruct()) / I32F32::from_num(Perbill::ACCURACY);
+        let y: I64F64 = floatpow(I32F32::from_num(1) + x, exponent)
+            .expect("Cannot overflow since rounds_per_year is u32 so worst case 0; QED");
```

Comment:
I counted `expect`s in this pallet, there're too many of `expect`.
Can some of them be removed?